src/main/java/rx/internal/operators/OnSubscribeSwitchIfEmpty.java

```diff
@@ -69,14 +82,33 @@ public void onCompleted() {
             if (!empty) {
                 child.onCompleted();
             } else if (!child.isUnsubscribed()) {
-                subscribeToAlternate();
+                active = false;
+                subscribe(null);
             }
         }
 
-        private void subscribeToAlternate() {
-            AlternateSubscriber<T> as = new AlternateSubscriber<T>(child, arbiter);
-            serial.set(as);
-            alternate.unsafeSubscribe(as);
+        void subscribe(Observable<? extends T> source) {
+            if (wip.getAndIncrement() == 0) {
+                do {
+                    if (child.isUnsubscribed()) {
+                        break;
+                    }
+
+                    if (!active) {
+                        if (secondary) {
```

Comment:
Unfortunately, no. Here is a scenario:

1) Thread 1 subscribes and calls `subscribe(source)`
2) through subscribeOn, Thread 2 calls `onCompleted()`, and thus `subscribe(null)` but finds Thread 1 still inside the loop (`wip != 0`).
3) Thread 1 detects the completion of Thread 2 and loops back. Since `source != null` still in this method, the main source is resubscribed again instead of the alternative.

The `secondary`, persisted indicator works across threads and makes sure no matter where onComplete comes from, the second time we get into that inner part, the alternative source is subscribed next.